core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeTransformationManager.java

```diff
@@ -104,37 +98,127 @@ public void run() {
-                message.append("\"event\":").append(messages.get(i));
-                message.append("}");
-                message.append(",");
-                totalBatchSize += Utils.getUTF8Length(message);
-                if (totalBatchSize >= Utils.MAX_BATCH_SIZE) {
-                    RudderLogger.logDebug(String.format(Locale.US, "DeviceModeTransformationManager: createDeviceTransformPayload: MAX_BATCH_SIZE reached at index: %d | Total: %d", i, totalBatchSize));
-                    break;
-                }
-                jsonPayload.append(message);
+    private void createMessageMap() {
+        for (int i = 0; i < messageIds.size(); i++) {
+            RudderMessage message = gson.fromJson(messages.get(i), RudderMessage.class);
+            RudderTransformationRequest transformationRequest = new RudderTransformationRequest();
+            transformationRequest.setMessage(message);
+            messageMap.put(messageIds.get(i), transformationRequest);
+        }
+    }
+
+    private boolean handleTransformationResponse(Result result, String requestJson) {
+        if (result.status == NetworkResponses.WRITE_KEY_ERROR) {
+            RudderLogger.logDebug("DeviceModeTransformationManager: TransformationProcessor: Wrong WriteKey. Aborting");
+            return true;
+        } else if (result.status == NetworkResponses.NETWORK_UNAVAILABLE) {
+            RudderLogger.logDebug("DeviceModeTransformationManager: TransformationProcessor: Network unavailable. Aborting");
+            return true;
+        } else if (result.status == NetworkResponses.ERROR) {
+            handleError(requestJson);
+        } else if (result.status == NetworkResponses.RESOURCE_NOT_FOUND) { // dumping back the original messages itself to the factories as transformation feature is not enabled
+            handleResourceNotFound(requestJson);
+        } else {
+            handleSuccess(result);
+        }
+        return false;
+    }
+
+    private void handleError(String requestJson) {
+        int delay = Math.min((1 << retryCount) * 500, MAX_DELAY); // Exponential backoff
+        if (retryCount++ == MAX_RETRIES) {
+            retryCount = 0;
+            deviceModeSleepCount = 0;
+            rudderDeviceModeManager.dumpOriginalEvents(gson.fromJson(requestJson, TransformationRequest.class), true);
+            completeDeviceModeEventProcessing();
+        } else {
+            RudderLogger.logDebug("DeviceModeTransformationManager: TransformationProcessor: Retrying in " + delay + "s");
+            try {
+                Thread.sleep(delay);
+            } catch (Exception e) {
+                RudderLogger.logError(e);
+                Thread.currentThread().interrupt();
             }
+        }
+    }
+
+    private void handleResourceNotFound(String requestJson) {
+        deviceModeSleepCount = 0;
+        rudderDeviceModeManager.dumpOriginalEvents(gson.fromJson(requestJson, TransformationRequest.class), false);
```

Comment:
Can't the messageMap be used to get the request?

Reply:
I've refactored the code further, please review it again.